cupy/cuda/memory.pyx

```diff
@@ -556,14 +559,42 @@ cpdef set_allocator(allocator=None):
         allocator = _malloc
     _current_allocator = allocator
 
-
 cpdef get_allocator():
     """Returns the current allocator for GPU memory.
 
     Returns:
         function: CuPy memory allocator.
     """
-    return _current_allocator
+    allocators = getattr(_thread_local, 'allocators', [_current_allocator])
+    return allocators[-1]
+
+
+@contextlib.contextmanager
+def using_allocator(allocator=None):
+    """Sets a thread-local allocator for GPU memory inside
+       context manager
+
+    Args:
+        allocator (function): CuPy memory allocator. It must have the same
+            interface as the :func:`cupy.cuda.alloc` function, which takes the
+            buffer size as an argument and returns the device buffer of that
+            size. When ``None`` is specified, raw memory allocator will be
+            used (i.e., memory pool is disabled).
+
+    """
```

Comment:
How about the following implementation for simplification?
```py
    if allocator is None:
        allocator = _malloc
    previous_allocators = getattr(_thread_local, 'allocator', None)
    _thread_local.allocators = allocator
    try:
        yield
    finally:
        _thread_local.allocator = previous_allocator
        if previous_allocator is None:
            delattr(_thread_local, 'allocator')
```

Reply:
Won't the current allocator get overwritten in nested context managers?

Reply:
Thanks for the suggestion, I should have realized it myself😢